base/ltmarks.dtx

```diff
@@ -987,29 +1118,160 @@
                     }
                 }
             }
-        }
+}
 %    \end{macrocode}
-%    If the badness was zero (we actually tested for${}>0$ but it
-%    can't get negative) then we had infinite shrinkage, so we report
-%    that and set all marks to the value the last mark had before.
+%  \end{macro}
+
+
+
+
+%  \begin{macro}{\@@_update_structure_when_infinite_glue:n}
+%    If a region contains infinite negative glue we report this as an
+%    error and then set all marks to the last mark from the previous
+%    region (as we can't extract any new marks).
 %    \begin{macrocode}
-        {
-          \msg_error:nnn { mark } { infinite-shrinkage } {#1}
-          \seq_map_inline:Nn \g_@@_classes_seq
+\cs_new_protected:Npn \@@_update_structure_when_infinite_glue:n #1 {
+  \msg_error:nnn { mark } { infinite-shrinkage-region } {#1}
+  \seq_map_inline:Nn \g_@@_classes_seq
+      {
+        \tl_gset_eq:cc { g_@@_#1_top_  ##1_tl }
+                       { g_@@_#1_last_ ##1_tl }
+        \tl_gset_eq:cc { g_@@_#1_first_##1_tl }
+                       { g_@@_#1_last_ ##1_tl }
+      }
+}
+%    \end{macrocode}
+%  \end{macro}
+%
+%
+%
+%  \begin{macro}{\@@_get_marks_for_reinsertion:nNN}
+%
+%    This function extracts the marks from the material in the first
+%    argument but it does not update any the mark structure. Instead,
```

Comment:
. . . update any mark structures. ?? 
OR:
. . . update the mark structure(s). ??